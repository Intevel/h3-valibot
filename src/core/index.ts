import {
  boolAsString,
  checkboxAsString,
  intAsString,
  numAsString,
  uuid,
} from './schemas'

export const vh = {
  boolAsString,
  checkboxAsString,
  intAsString,
  numAsString,
  uuid,
}

export {
  useSafeValidatedBody,
  useSafeValidatedParams,
  useSafeValidatedQuery,
  useValidatedBody,
  useValidatedParams,
  useValidatedQuery,
  type VSchema,
} from './parsers'

export * as v from 'valibot'
