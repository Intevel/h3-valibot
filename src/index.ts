export { useValidatedBody, useValidatedParams, useValidatedQuery } from './parsers';
