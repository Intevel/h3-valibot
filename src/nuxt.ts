import { addServerImports, createResolver, defineNuxtModule } from '@nuxt/kit'
import '@nuxt/schema'

export default defineNuxtModule({
  meta: {
    name: 'h3-valibot',
  },
  defaults: {},
  async setup(_options, _nuxt) {
    const { resolve } = createResolver(import.meta.url)
    addServerImports([
      {
        name: 'useSafeValidatedBody',
        from: resolve('./core'),
      },
      {
        name: 'useSafeValidatedParams',
        from: resolve('./core'),
      },
      {
        name: 'useSafeValidatedQuery',
        from: resolve('./core'),
      },
      {
        name: 'useValidatedBody',
        from: resolve('./core'),
      },
      {
        name: 'useValidatedParams',
        from: resolve('./core'),
      },
      {
        name: 'useValidatedQuery',
        from: resolve('./core'),
      },
    ])
    addServerImports([
      {
        name: 'v',
        from: resolve('./core'),
      },
      {
        name: 'vh',
        from: resolve('./core'),
      },
    ])
  },
})
