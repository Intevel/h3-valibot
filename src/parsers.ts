import { createError, getQuery, getRouterParams, type H3Event, readBody } from 'h3'
import * as v from 'valibot'

type VSchema<TInput, TOutput, TIssue extends v.BaseIssue<unknown>> =
  | v.BaseSchema<TInput, TOutput, TIssue>
  | v.BaseSchemaAsync<TInput, TOutput, TIssue>

const DEFAULT_ERROR_MESSAGE = 'Bad Request'
const DEFAULT_ERROR_STATUS = 400

function createBadRequest(error: any) {
  return createError({
    statusCode: DEFAULT_ERROR_STATUS,
    statusText: DEFAULT_ERROR_MESSAGE,
    data: error,
  })
}

/**
 * Parse and validate request query from event handler. Throws an error if validation fails.
 * @param event - A H3 event object.
 * @param schema - A Valibot Schema
 */
export async function useValidatedQuery<
  TInput,
  TOutput,
  TIssue extends v.BaseIssue<unknown>,
>(
  event: H3Event,
  schema: VSchema<TInput, TOutput, TIssue>,
): Promise<TOutput> {
  try {
    const query = getQuery(event)
    const parsed = await v.parseAsync(schema, query)
    return parsed
  }
  catch (error) {
    throw createBadRequest(error)
  }
}

/**
 * Parse and validate request query from event handler. Doesn't throw if validation fails.
 * @param event - A H3 event object.
 * @param schema - A Valibot Schema
 */
export async function useSafeValidatedQuery<
  TInput,
  TOutput,
  TIssue extends v.BaseIssue<unknown>,
>(
  event: H3Event,
  schema: VSchema<TInput, TOutput, TIssue>,
): Promise<v.SafeParseResult<VSchema<TInput, TOutput, TIssue>>> {
  try {
    const query = getQuery(event)
    const parsed = await v.safeParseAsync(schema, query)
    return parsed
  }
  catch (error) {
    throw createBadRequest(error)
  }
}

/**
 * Parse and validate request body from event handler. Throws an error if validation fails.
 * @param event - A H3 event object.
 * @param schema - A Valibot Schema
 */
export async function useValidatedBody<
  TInput,
  TOutput,
  TIssue extends v.BaseIssue<unknown>,
>(
  event: H3Event,
  schema: VSchema<TInput, TOutput, TIssue>,
): Promise<TOutput> {
  try {
    const body = await readBody(event)
    const parsed = await v.parseAsync(schema, body)
    return parsed
  }
  catch (error) {
    throw createBadRequest(error)
  }
}

/**
 * Parse and validate request body from event handler. Doesn't throw if validation fails.
 * @param event - A H3 event object.
 * @param schema - A Valibot Schema
 */
export async function useSafeValidatedBody<
  TInput,
  TOutput,
  TIssue extends v.BaseIssue<unknown>,
>(
  event: H3Event,
  schema: VSchema<TInput, TOutput, TIssue>,
): Promise<v.SafeParseResult<VSchema<TInput, TOutput, TIssue>>> {
  try {
    const body = await readBody(event)
    const parsed = await v.safeParseAsync(schema, body)
    return parsed
  }
  catch (error) {
    throw createBadRequest(error)
  }
}

/**
 * Parse and validate request params from event handler. Throws an error if validation fails.
 * @param event - A H3 event object.
 * @param schema - A Valibot Schema
 */
export async function useValidatedParams<
  TInput,
  TOutput,
  TIssue extends v.BaseIssue<unknown>,
>(
  event: H3Event,
  schema: VSchema<TInput, TOutput, TIssue>,
): Promise<TOutput> {
  try {
    const params = getRouterParams(event)
    const parsed = await v.parseAsync(schema, params)
    return parsed
  }
  catch (error) {
    throw createBadRequest(error)
  }
}

/**
 * Parse and validate request params from event handler. Doesn't throw if validation fails.
 * @param event - A H3 event object.
 * @param schema - A Valibot Schema
 */
export async function useSafeValidatedParams<
  TInput,
  TOutput,
  TIssue extends v.BaseIssue<unknown>,
>(
  event: H3Event,
  schema: VSchema<TInput, TOutput, TIssue>,
): Promise<v.SafeParseResult<VSchema<TInput, TOutput, TIssue>>> {
  try {
    const params = getRouterParams(event)
    const parsed = await v.safeParseAsync(schema, params)
    return parsed
  }
  catch (error) {
    throw createBadRequest(error)
  }
}
