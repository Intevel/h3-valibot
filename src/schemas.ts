// Original code by https://github.com/rileytomasek/zodix/blob/31bd5f2708f2eec4d0522201710dfc5ef887a16e/src/schemas.ts
import * as v from 'Valibot'

/**
 * Valibot schema to parse strings that are booleans.
 * Use to parse <input type="hidden" value="true" /> values.
 * @example
 * ```ts
 * v.parse(boolAsString, 'true') -> true
 * ```
 */
export const boolAsString = v.pipe(
  v.string(),
  v.regex(/^(true|false)$/, 'Must be a boolean string ("true" or "false")'),
  v.transform(value => value === 'true'),
)

/**
 * Valibot schema to parse checkbox formdata.
 * Use to parse <input type="checkbox" /> values.
 * @example
 * ```ts
 * v.parse(checkboxAsString, 'on') -> true
 * v.parse(checkboxAsString, undefined) -> false
 * ```
 */
export const checkboxAsString = v.optional(v.pipe(
  v.picklist(['on', 'off']),
  v.transform(value => value === 'on'),
), 'off')

/**
 * Valibot schema to parse strings that are integers.
 * Use to parse  <input type="number" /> values.
 * @example
 * ```ts
 * v.parse(intAsString, '3') -> 3
 * ```
 */
export const intAsString = v.pipe(
  v.string(),
  v.regex(/^-?\d+$/, 'Must be an integer string'),
  v.transform(val => Number.parseInt(val, 10)),
)

/**
 * Valibot schema to parse strings that are numbers.
 * Use to parse <input type="number" step="0.1" /> values.
 * @example
 * ```ts
 * v.parse(numAsString, '3.14') -> 3.14
 * ```
 */
export const numAsString = v.pipe(
  v.string(),
  v.regex(/^-?(?:\d+(?:\.\d+)?|\.\d+)$/, 'Must be a number string'),
  v.transform(Number),
)
