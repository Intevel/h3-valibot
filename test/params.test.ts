import type { App, Router } from 'h3'
import type { Test } from 'supertest'
import type TestAgent from 'supertest/lib/agent'
import { createApp, createRouter, defineEventHandler, toNodeListener } from 'h3'
import supertest from 'supertest'
import { beforeEach, describe, expect, it } from 'vitest'
import { useSafeValidatedParams, useValidatedParams, v } from '../src'

describe('useValidatedParams', () => {
  let app: App
  let router: Router
  let request: TestAgent<Test>

  beforeEach(() => {
    app = createApp({ debug: false })
    router = createRouter()
    app.use(router)
    request = supertest(toNodeListener(app))
  })

  const paramsSchema = v.object({
    name: v.string(),
  })

  it('returns 200 OK if params matches validation schema', async () => {
    router.get('/validate/:name', defineEventHandler(event => useValidatedParams(event, paramsSchema)))

    const res = await request.get('/validate/sandros94')

    expect(res.status).toEqual(200)
    expect(res.body).toMatchSnapshot()
  })

  it('throws 400 Bad Request if params does not match validation schema', async () => {
    router.get('/validate/:name', defineEventHandler(event => useValidatedParams(event, paramsSchema)))

    const res = await request.get('/validate')

    expect(res.status).toEqual(404)
    expect(res.body).toMatchSnapshot()
  })

  it('doesn\'t throw 400 Bad Request if params does not match validation schema', async () => {
    router.get('/validate/:name', defineEventHandler(event => useSafeValidatedParams(event, paramsSchema)))

    const res = await request.get('/validate/2') // TODO: this is not actually correct

    expect(res.status).toEqual(200)
    expect(res.body).toMatchSnapshot()
  })
})
