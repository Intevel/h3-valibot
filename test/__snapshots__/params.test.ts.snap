// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useValidatedParams > doesn't throw 400 Bad Request if params does not match validation schema 1`] = `
{
  "output": {
    "name": "2",
  },
  "success": true,
  "typed": true,
}
`;

exports[`useValidatedParams > returns 200 OK if params matches validation schema 1`] = `
{
  "name": "sandros94",
}
`;

exports[`useValidatedParams > throws 400 Bad Request if params does not match validation schema 1`] = `
{
  "stack": [],
  "statusCode": 404,
  "statusMessage": "Cannot find any path matching /validate.",
}
`;
